### ISSUE TITLE
[kmac] D2S Signoff

### ISSUE THREAD

**Issue by msfschaffner** [lowRISC/opentitan#20978]
### Description

Ensure D2S signoff criteria are fulfilled after focus area changes have landed.

**Comment by vogelpi** [comment]
## Commits since Earlgrey-ES tapeout
```
git rev-parse --short HEAD
```
3d21fe76c5
```
git log Earlgrey-M2.5.2-RC0..HEAD --oneline hw/ip/kmac
```
15f75fa343 [kmac, sha3] Add REQ/ACK interface to delay Keccak operations
5cb6b899d7 [kmac] Rework fifo_empty interrupt, change type to status
adf455da16 [kmac] Add separate FuseSoc core file for kmac_reduced.sv
ddc64ed3c2 [kmac] Fix driving conflict in kmac_reduced.sv used for SCA evaluation
642d7a9733 [kmac] Switch to Trivium-based PRNG implementation
4414aaab56 [kmac] Add param to keccak_round/2share to only use external randomness
0237c23809 [kmac] Add buffer stage to PRNG output
365dfd2ba9 [kmac] Make DOM multiplier I/O muxing glitch free
dd4b689489 [kmac] Move DOM multiplier control from keccak_2share to keccak_round
043b0e0783 [kmac] Simplify randomness update requests and PRNG control logic
fa5dc8a876 [pre_sca] Convert PROLEAD configuration files to Unix format
deb7ee190f [kmac,pre_dv] Check digests produced by kmac_reduced_tb using DPI model
72c05a2c8d [kmac] Add missing CM annotation
178e0883b9 [kmac, pre_sca] Add PROLEAD setup
b56df44065 [kmac,pre_dv] Add scratch Verilator testbench for kmac_reduced module
bf71a0f085 [kmac,rtl] Add kmac_reduced module suitable for SCA e.g. using PROLEAD
66472e257a [pre_syn] Include csrng_pkg.sv to re-enable Yosys synthesis
56399241e7 Revert "[edn] Move prim_edn_req out of prim"
c721c51c13 [rtl, prim] Add 'commit' functionality to prim_count
61a237e197 [util/reggen] reverse order of substruct generation
fc8484601e [reggen,hw] Create index parameter for registers windows
3b4e36e01c [edn] Move prim_edn_req out of prim
de31bdf1c2 [reggen] Remove the devmode input
963a5006cc [doc] Minor tweak to md sanitisation code
5be278bb25 [aes, kmac, otbn] Perform final `clean -purge` step in Yosys synthesis
2257a7276f [kmac/pre_syn] Re-enable Yosys synthesis flow
975a6eb927 [adc_ctrl,dv] Tidy up access to intr_state in env_cfg files
1b16ca2122 [reggen] Add mubi support SWAccess that sets/clears a reg
b968a168c2 [kmac,doc] Correct performance numbers
f96f3dc13e [kmac, doc] Update doc with cmdgen.py
3c02ff2603 [kmac, doc] Update `MSG_FIFO` description
59f8142826 [doc] Moved badges over to using hosted images
9efb9cc879 [doc] kmac registers and interfaces now use CMDGEN
8ef4c5525d [chip-test] SiVal test plan updates for KMAC
7688e714e8 [reggen] Add initial support for version and cip_id hjson fields
fbd888eea8 Revert "[reggen] Add CIP_IDs and bump all major versions"
2db7ac46d8 [kmac, rtl] Advance LFSR upon randomness consumption during reseeding
e53fe26466 [kmac, rtl] Tweak SVA checking that the Keccak core only uses valdi PRD
37314cdf62 [kmac, doc] Fix broken references
9bc003ca25 [aes, kmac] Replace term aggravate in SCA/FI context
b0044a4753 [kmac, rtl] Hold entropy_req until it's acknowledged
b156e6d0cc [kmac, rtl] Keep advancing LFSR in StRandErr as during normal operation
0ba10b3cd3 [reggen] Add CIP_IDs and bump all major versions
e47df29f3e [misc] Use lc_tx_t testing functions at endpoints

## Issues closed since the Earlgrey-ES tapeout

### DD & DV
- #21049
  Changing the previously unused `fifo_empty` interrupt from type `Event` to `Status`. The interrupt isn't currently verified at the block level and also wasn't before this change.
- #20828 
  The main design change for Earlgrey-PROD. As for AES, the PRNG design has been switched to a Trivium-based implementation to avoid potential brute forcing vulnerabilities on the PRNG state (previously made up by 25 32-bit LFSRs). While doing this work, the controls of the PRNG as well as the control of the masked Keccak core have been optimized to improve the SCA hardening. A full security evaluation has been performed before merging the change.
- #19232
  Two bugs in the handling of EDN requests (1. dropping the request without ack, 2. not advancing the PRNG if EDN times out). Both have been fixed.
- #14596 
  A duplicate of #16855 which is tracked for M3.  

### Doc
- #20887 
- #17256 

### SW
- #20674
  Some changes to the SCA setup to enable measurements using the actual ES silicon.

### SiVal / Chip-level tests
- #21436
  Chip-level test issue, duplicate of #16069 .
- #21403 
  Chip-level test issue, duplicate of #20443 .
- #20443
  Chip-level test issue, the test was already implemented but not yet cross referenced.
  
### Misc
- #12323
  A couple of follow-up tasks. Fanned out into separate issues where appropriate, provided arguments of what can be closed / ignored.

## Currently open issues

### DD / DV
- #10724 
  A minor RTL fix / security improvement in the ROM_CTRL / KMAC interface. Uncontroversial and tracked for M3.
- #10704 
  A small security improvement tracked for implementation in M3. No big area of physical design impact.
- #16855
  Strongly related to #10704 if not a duplicate. small security improvement. No PD impact, tracked for M3.
- #17209
  Moving the message masking to after the padding to improve SCA hardening. No PD impact. Now that the masking is improved, most likely no longer needed. M3.
- #16202
  No work for KMAC needed as the KMAC S-Box is super simple compared to the AES one. 
- #3479
  Context save and restore. Not under discussion for implementation currently.
- #18029 
  For Keccak state sizes other than what is used for Earlgrey and Darjeeling, a width mismatch could occur inside `keccak_round.sv`. This will most likely generate a lint error and simulation stops due to an SVA. This can be fixed if it really turns out it's needed sometime in the future.
- #17585
  A security item tracked for keymgr (M3).

### DV
- #17759
  A V3 item (M5)
#16269
  A V2S item (M5)
- #18644
  A failing block-level test, should be investigated as part of V3 closure (M5)

### Doc
- #17785 
  Entropy-related error conditions inside KMAC are not reported to keymgr (see #15088). This issue is about formulating some guidance for software developers to check for those error conditions inside KMAC when using keymgr.
- #15088
  See above. 
- #16792
  Formulating guidance for software developers. 

### SiVal / Chip-level tests
- #21413 
- #21412 
- #21411 
- #21410 
- #21409 
- #21408 
- #21407 
- #21406 
- #21405 
- #21404 

### Misc
- #6543
  A penetration testing item.

## Summary

Since Earlgrey-M2.5.2-RC0, there have been two notable RTL changes to KMAC (besides some minor bug fixes):
- The change of the PRNG (together with optimizing the masking SCA countermeasure)
- The change of the interrupt type for the `fifo_empty` interrupt
Both changes are reflected in the documentation and DV has been aligned where possible. A full security evaluation has been performed for the masking / PRNG change. The `fifo_empty` interrupt isn't verified (but it wasn't before the change) and software currently does not use it. Based on this, I feel confident to declare D2S.

For M3, there three minor security improvements outstanding (see open issues). But none of this is going to be a major change. Thus, I am still in favor of declaring D2S (the block was D2S before) but I'd also be okay to go to D2 first. PLMK @msfschaffner , @andreaskurth .

**Comment by andreaskurth** [comment]
I agree that D2S is still fulfilled. I reviewed the following commits, which change `kmac`'s RTL code (commit list obtained with `git log --oneline Earlgrey-M2.5.2-RC0..HEAD -- hw/ip/kmac/rtl`):
- 15f75fa343 [kmac, sha3] Add REQ/ACK interface to delay Keccak operations
- 5cb6b899d7 [kmac] Rework fifo_empty interrupt, change type to status
- ddc64ed3c2 [kmac] Fix driving conflict in kmac_reduced.sv used for SCA evaluation
- 642d7a9733 [kmac] Switch to Trivium-based PRNG implementation
- 4414aaab56 [kmac] Add param to keccak_round/2share to only use external randomness
- 0237c23809 [kmac] Add buffer stage to PRNG output
- 365dfd2ba9 [kmac] Make DOM multiplier I/O muxing glitch free
- dd4b689489 [kmac] Move DOM multiplier control from keccak_2share to keccak_round
- 043b0e0783 [kmac] Simplify randomness update requests and PRNG control logic
- 72c05a2c8d [kmac] Add missing CM annotation
- bf71a0f085 [kmac,rtl] Add kmac_reduced module suitable for SCA e.g. using PROLEAD
- c721c51c13 [rtl, prim] Add 'commit' functionality to prim_count
- 61a237e197 [util/reggen] reverse order of substruct generation
- fc8484601e [reggen,hw] Create index parameter for registers windows
- de31bdf1c2 [reggen] Remove the devmode input
- 1b16ca2122 [reggen] Add mubi support SWAccess that sets/clears a reg
- fbd888eea8 Revert "[reggen] Add CIP_IDs and bump all major versions"
- 2db7ac46d8 [kmac, rtl] Advance LFSR upon randomness consumption during reseeding
- e53fe26466 [kmac, rtl] Tweak SVA checking that the Keccak core only uses valdi PRD
- 9bc003ca25 [aes, kmac] Replace term aggravate in SCA/FI context
- b0044a4753 [kmac, rtl] Hold entropy_req until it's acknowledged
- b156e6d0cc [kmac, rtl] Keep advancing LFSR in StRandErr as during normal operation
- 0ba10b3cd3 [reggen] Add CIP_IDs and bump all major versions
- e47df29f3e [misc] Use lc_tx_t testing functions at endpoints

The main RTL changes are in the following PRs:
- https://github.com/lowRISC/opentitan/pull/19248
- https://github.com/lowRISC/opentitan/pull/19262
- https://github.com/lowRISC/opentitan/pull/21624
- https://github.com/lowRISC/opentitan/pull/21657
- https://github.com/lowRISC/opentitan/pull/21787

The first three PRs have been thoroughly reviewed, resulting in approval by multiple committers. I scanned the changes again and did not notice problems that would prevent D2S sign-off in my view. The bottom two PRs have not been scrutinized to the same degree, so I reviewed them again in detail and didn't find problems there either. With the security evaluation done as part of #21624, we should be able to maintain the maturity level of the countermeasures.

Thus I think we should proceed with signing KMAC off at D2S.

**Comment by andreaskurth** [comment]
I justed noticed that KMAC is still at D2S and didn't have its version number increased:
https://github.com/lowRISC/opentitan/blob/95d1385c70b737042376a4b53dda8c907e0ff449/hw/ip/kmac/data/kmac.hjson#L27-L32

Given the PRNG reworking, the type change of the `fifo_empty` interrupt, and the fixing of the CS AES Halt interface, I think we should bump at least the minor version (i.e., `1.1.0`) (and remove the outdated `commit_id`). WDYT @vogelpi @msfschaffner ?

**Comment by andreaskurth** [comment]
As just discussed in the WG meeting, @msfschaffner recommends a major version bump due to the SW-visible interface change (interrupt type), and I agree.

**Comment by vogelpi** [comment]
Speaking of SW-visible interface, I've now realized that already the PRNG change (#21624) was software visible because KMAC features a way for SW to reseed the PRNG initially. Since the state size of the PRNG changed, the number of words to be written needed to be changed and to streamline things, I've switched from 5 registers written once each to 1 register written 9 times. So, the major version bump definitely makes sense to me.
